alacritty/src/input.rs

```diff
@@ -492,26 +492,77 @@ impl<'a, T: EventListener, A: ActionContext<T>> Processor<'a, T, A> {
             };
 
             self.mouse_report(code, ElementState::Pressed);
-        } else if button == MouseButton::Left {
-            self.on_left_click();
         } else {
-            // Do nothing when using buttons other than LMB.
-            self.ctx.mouse_mut().click_state = ClickState::None;
+            // Reset click state if button has changed
+            if button != self.ctx.mouse().last_click_button {
+                self.ctx.mouse_mut().last_click_button = button;
+                self.ctx.mouse_mut().click_state = ClickState::None;
+            }
+
+            // Calculate time since the last click to handle double/triple clicks.
+            let now = Instant::now();
+            let elapsed = now - self.ctx.mouse().last_click_timestamp;
+            self.ctx.mouse_mut().last_click_timestamp = now;
+
+            // Load mouse point, treating message bar and padding as closest cell.
+            let mouse = self.ctx.mouse();
+            let mut point = self.ctx.size_info().pixels_to_coords(mouse.x, mouse.y);
+            point.line = min(point.line, self.ctx.terminal().grid().num_lines() - 1);
+
+            match button {
+                MouseButton::Left => self.on_left_click(point, elapsed),
+                MouseButton::Right => self.on_right_click(point, elapsed),
+                // Do nothing when using buttons other than LMB.
+                _ => self.ctx.mouse_mut().click_state = ClickState::None,
+            }
         }
     }
 
-    /// Handle left click selection and vi mode cursor movement.
-    fn on_left_click(&mut self) {
-        // Calculate time since the last click to handle double/triple clicks in normal mode.
-        let now = Instant::now();
-        let elapsed = now - self.ctx.mouse().last_click_timestamp;
-        self.ctx.mouse_mut().last_click_timestamp = now;
+    /// Handle selection expansion on right click.
+    fn on_right_click(&mut self, point: Point, elapsed: Duration) {
+        self.ctx.mouse_mut().click_state = match self.ctx.mouse().click_state {
```

Comment:
I feel like we can compute click state earlier and use proper click state later? In general it could perform click state computation before `on_left_click`/`on_right_click` and then just match current click state in those handlers.

Reply:
That's not a bad idea actually. I'll look into it.

Reply:
Of course this requires matching twice, but I think it should be worth it for the sake of cleaner code.